cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -293,74 +289,46 @@ private void inspectAdditionalPlugins(ArtifactId artifactId, Set<PluginClass> ad
-                Enumeration<URL> resources = pluginClassLoader.getResources(resourceName);
-                List<Iterator<String>> iterators = new ArrayList<>();
-                // Go though all available resources and collect all class names that are plugin classes.
-                while (resources.hasMoreElements()) {
-                  URL packageResource = resources.nextElement();
-
-                  // Only inspect classes in the top level jar file for Plugins.
-                  // The jar manifest may have packages in Export-Package that are loadable from the bundled jar files,
-                  // which is for classloading purpose. Those classes won't be inspected for plugin classes.
-                  // There should be exactly one of resource that match, because it maps to a directory on the FS.
-                  if (packageResource.getProtocol().equals("file")) {
-                    Iterator<String> classFiles = DirUtils.list(new File(packageResource.toURI()), "class").iterator();
-
-                    // Transform class file into class name and filter by @Plugin class only
-                    iterators.add(Iterators.filter(
-                      Iterators.transform(classFiles, new Function<String, String>() {
-                        @Override
-                        public String apply(String input) {
-                          return getClassName(currentPackage, input);
-                        }
-                      }), new Predicate<String>() {
-                        @Override
-                        public boolean apply(String className) {
-                          return isPlugin(className, pluginClassLoader);
-                        }
-                      }));
-                  }
-                }
-                if (!iterators.isEmpty()) {
-                  classIterator = Iterators.concat(iterators.iterator());
-                }
-              } catch (Exception e) {
-                // Cannot happen
-                throw Throwables.propagate(e);
-              }
-            }
-
-            try {
-              return pluginClassLoader.loadClass(classIterator.next());
-            } catch (ClassNotFoundException | NoClassDefFoundError e) {
-              // Cannot happen, since the class name is from the list of the class files under the classloader.
-              throw Throwables.propagate(e);
-            }
+  private Iterable<Class<?>> getPluginClasses(Collection<String> packages,
+                                              PluginClassLoader pluginClassLoader) {
+    Predicate<String> nameCheckPredicate = getClassNameCheckPredicate(packages);
+    try (JarFile jarFile = new JarFile(pluginClassLoader.getTopLevelJar())) {
+      return jarFile
+        .stream()
+        .filter(entry -> !entry.isDirectory())
+        .map(ZipEntry::getName)
+        .filter(nameCheckPredicate)
+        .map(fileName -> StringUtils.removeEnd(fileName, ".class").replace('/', '.'))
```

Comment:
If possible, I would avoid using the apache common utils (with a long term goal to reduce dependencies) as long as it is still maintained as a one-liner (e.g. `fileName.substring(0, fileName.length() - ".class".length().replace('/',' .')`  in this case since we know the name is ended with ".class").

Reply:
Done